### PR TITLE
Uncached docker build

### PR DESCRIPTION
**Pull request by steveri**
### Executive summary
1. Docker builds no longer run out of memory.
2. ~Docker image size is now less than 8GB, down from the previous 12GB.~

### Details
1. Okay I found the secret to keeping the builds from running out of memory, and that is to turn off caching (see changes in file `dockerimage.yml`).  Apparently, the docker build has an option to cache intermediate results such that future builds can reuse that information rather than having to build it from scratch. Of course, the cache requires disk space big enough for the cached information, and our image build apparently filled that up...

    The original recommendation seems to have been that you should turn on caching for every git push, so that later git-push builds could proceed more quickly. Meanwhile, scheduled daily cron-job builds would do uncached builds, I guess because these jobs are presumably not producing any information that's not already in the cache.

    Because our builds happen in the cloud, on virtual machines that get built on demand just for that build, the caching seems like it doesn't make much sense anyway...won't the cached information just go away when the build is finished and the VM is shut down?

    Anyway! I turned off caching and magically everything started working again, so apparently that was the right thing to do.

~2. I rejiggered Dockerfile to produce fewer layers and to delete intermediate files and such before saving each layer, so as to produce a smaller final image size. Previously, images were 12GB, now they are 8GB. Below we can see the result of a recent trial on kiwi (`latest` is the most recent successful master-branch build, `uncached-docker-build` is the one produced by this new branch):~

Did not do this (yet):
```
% docker images
stanfordaha/garnet   uncached-docker-build   724b059ef9ea   6 minutes ago   7.89GB
stanfordaha/garnet   latest                  c7a5e18ac6fd   2 days ago      11.6GB
```

### CODE REVIEW

**Comment by kalhankoul96** [comment]
This seems good, this is different from your original idea of not deiniting right? In that case this is great because it won't change anyones workflow.

**Review by kalhankoul96 — APPROVED** [review]

**Comment by kalhankoul96** [comment]
Looks like you dropped some of the pono/smt-switch install lines.

File: Dockerfile [hunk 1]
```diff
@@ -61,21 +61,39 @@ RUN apt-get update && \
...
-WORKDIR /aha/pono/build
-RUN make -j4 && pip install -e ./python
-WORKDIR /aha
+RUN pip install Cython==0.29 pytest toml scikit-build==0.13.0 && \
+  ./contrib/setup-bison.sh && ./contrib/setup-flex.sh && \
+  ./contrib/setup-smt-switch.sh --python && ./contrib/setup-btor2tools.sh && \
+  cd /aha/pono && ./configure.sh --python && \
+  cd /aha/pono/build && make -j4 && pip install -e ./python && \
+  source /aha/bin/activate && \
+  pip install -e /aha/pono/deps/smt-switch/build/python && \
```
Comment (steveri) [inline Dockerfile:79]: Moved pono pip-install to here, so we can clean up afterward...

**Comment by steveri** [comment]
> Looks like you dropped some of the pono/smt-switch install lines.

Good catch! But no, they are not dropped. I moved them back (see line 79 ish), closer to where pono is first initiated, which seems like where they should have been in the first place. We can talk more about this when we meet today...

**Comment by steveri** [comment]
> This seems good, this is different from your original idea of deiniting right? In that case this is great because it won't change anyones workflow.

Haha my initial idea did not work AT ALL, not even remotely. I will tell you more when we meet.

**Comment by steveri** [comment]
> > Looks like you dropped some of the pono/smt-switch install lines.
> 
> Good catch! But no, they are not dropped. I moved them back (see line 79 ish), closer to where pono is first initiated, which seems like where they should have been in the first place. We can talk more about this when we meet today...

Huh maybe I'm not so clever after all. Looks like regressions are failing because of missing smt-switch. I will investigate.

**Comment by steveri** [comment]
Just talked to Kalhan, we agreed to remove the Dockerfile changes from this pull so that we can get docker images built and working ASAP. So for now the images will still be 12GB, but, with caching turned off, at least they should build without running out of disk space. The image-size reduction project will reappear in a future pull request.

**Comment by steveri** [comment]
Also: I will wait for the CI tests to complete successfully and then finalize the pull unless there are objections between now and then.

**Pull request merged**